plugins/http_plugin/beast_http_listener.hpp

```diff
@@ -0,0 +1,140 @@
+#pragma once
+
+#include "beast_http_session.hpp"
+#include "common.hpp"
+#include <type_traits>
+#include <sstream>
+
+namespace eosio {
+    // since beast_http_listener handles both TCP and UNIX endpoints we need a template here 
+    // to get the path if makes sense, so that we can call ::unlink() before opening socket 
+    // in beast_http_listener::listen() by tdefault return blank string
+    template<typename T>
+    std::string get_endpoint_path(T endpt) { return {}; }
+
+    std::string get_endpoint_path(stream_protocol::endpoint endpt) { return endpt.path(); }
+
+    // Accepts incoming connections and launches the sessions
+    // session_type should be a subclass of beast_http_session
+    // protocol type must have sub types acceptor and endpoint, e.g. boost::asio::ip::tcp;
+    // socket type must be the socket e.g, boost::asio::ip::tcp::socket
+    template<typename session_type, typename protocol_type, typename socket_type> 
+    class beast_http_listener : public std::enable_shared_from_this<beast_http_listener<session_type, protocol_type, socket_type> >
+    {
+        private: 
+            bool is_listening_ = false; 
+
+            std::shared_ptr<eosio::chain::named_thread_pool> thread_pool_;
+            std::shared_ptr<ssl::context> ctx_;
+            std::shared_ptr<http_plugin_state> plugin_state_;
+
+            typename protocol_type::acceptor acceptor_;
+            socket_type socket_;
+
+        public:
+            beast_http_listener() = default;
+            beast_http_listener(const beast_http_listener&) = delete;
+            beast_http_listener(beast_http_listener&&) = delete;
+
+            beast_http_listener& operator=(const beast_http_listener&) = delete;
+            beast_http_listener& operator=(beast_http_listener&&) = delete;
+
+            beast_http_listener(std::shared_ptr<eosio::chain::named_thread_pool> thread_pool, 
+                                std::shared_ptr<ssl::context> ctx, 
+                                std::shared_ptr<http_plugin_state> plugin_state) : 
+                is_listening_(false)
+                , thread_pool_(std::move(thread_pool))
+                , ctx_(std::move(ctx))
+                , plugin_state_(std::move(plugin_state))
+                , acceptor_(thread_pool_->get_executor())
+                , socket_(thread_pool_->get_executor())
```

Comment:
Even if we have convinced ourselves that an explicit strand is not required for our use case, I think it makes sense to still use a strand so that we more closely align with the boost beast examples.

Reply:
This is possible, although for boost versions prior to 1.70 the last solution I used was rather ugly.  If we do this I would prefer to bump the min boost version up to 1.70

The use of strands reinforces that the session objects are not thread-safe - i.e the object is intended to live, be accessed and die within a single thread.  Only the plugin_state objects are thread safe.  This is fine as we are not doing anything fancy like pipelining requests within a session.  There may be a slight perf penalty for using strands, though the vast majority of time (> 90%) is spent handling the request anyway.

Reply:
I would prefer not making it ugly. I've not been able to make it crash by bombarding it with requests, so I would prefer to leave as is for now.